BUG: Do not elide complex abs()

juliantaylor opened the pull request:
complex abs() results in float so it cannot be elided.

Closes gh-9109

juliantaylor committed:
BUG: do not elide complex abs()

complex abs() results in float so it cannot be elided.

Closes gh-9109

juliantaylor commented:
as far as I can tell abs() the only special unary ufunc that changes types

charris commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -2850,6 +2850,9 @@ def test_temporary_with_cast(self):
         d = f.astype(np.float64)
         assert_equal(((f + f) + d).dtype, np.dtype('f8'))
 
+        c = np.ones(100000, dtype=np.complex)
```
Might add note as to why large size is needed to test elision.

juliantaylor replied:
might make sense to move the elide test into its own class, its currently in binop but abs is a unary

eric-wieser commented on numpy/core/src/multiarray/number.c:
```diff
@@ -565,7 +565,7 @@ array_negative(PyArrayObject *m1)
 array_absolute(PyArrayObject *m1)
 {
-    if (can_elide_temp_unary(m1)) {
+    if (can_elide_temp_unary(m1) && !PyArray_ISCOMPLEX(m1)) {
```
Isn't this also going to fail for any user type that overrides `abs` too? #4730 comes to mind, and problems with a `quaternion` type.

juliantaylor replied:
hm right, it shouldn't elide object arrays at all it only does that for unary not bianry probably I forgot to update both places.

juliantaylor replied:
though abs(object-array) is again an object array regardless of what `__abs__` does

eric-wieser replied:
I'm referring to registering custom ufunc loops for new `dtype`s

eric-wieser replied:
Issue number should have been #9003

eric-wieser commented:
In general, it's not safe to elide any op on a user-defined dtype

juliantaylor committed:
BUG: only elide basic dtypes in unary ops

As in binary operations, only elide unary operations on basic types to
avoid odd interactions e.g. with user dtypes.

juliantaylor committed:
TST: move elision tests to their own class and add some more

juliantaylor commented:
disabled elision for all non-basic types for unary ops, they were already disable for binaries.
Also added some longdouble tests and some cleanup.

eric-wieser commented:
Why are object arrays not safe to elide?

juliantaylor commented:
they probably are, but it is probably not worth the time thinking about it as they are likely to slow to profit anyway.

charris commented:
@eric-wieser Are you OK with this?

eric-wieser commented on numpy/core/src/multiarray/temp_elide.c:
```diff
@@ -359,7 +359,7 @@ can_elide_temp_unary(PyArrayObject * m1)
     int cannot;
     if (Py_REFCNT(m1) != 1 || !PyArray_CheckExact(m1) ||
-            PyArray_DESCR(m1)->type_num == NPY_VOID ||
+            PyArray_DESCR(m1)->type_num >= NPY_OBJECT ||
```
Maybe just `!PyTypeNum_ISNUMBER(m)`? Do we want to include `NPY_HALF` here too?

juliantaylor replied:
the ee->f loop is only used for non fast powers which is not elided (though it could be, albeit with not much performance gain as normal pow is very slow)
adding a test for that is a good idea though

juliantaylor replied:
also I think that just defines the inner ufunc loop, the output buffer is actually still float16 so it can be elided.

juliantaylor replied:
it is so, no test needed then
using ISNUMBER is much better, added that

eric-wieser replied:
Should probably fix the release notes to say that elision is added for builtin _numeric_ types

juliantaylor committed:
MAINT: use PyArray_ISNUMBER to decide whether to elide

juliantaylor committed:
DOC: add numeric type to temp elision requirements

[ci skip]

eric-wieser reviewed (approved):
LGTM

charris commented:
Thanks Julian. And thanks Eric for reviewin, and feel free to do the merge without waiting on me.
